src/scoring/hpf_scoring.py

```diff
@@ -0,0 +1,168 @@
+        """Prediction function.
+
+        :param input_stack: The user's package list
+        for which companion recommendation are to be generated.
+        :return companion_recommendation: The list of recommended companion packages
+        along with condifence score.
+        :return package_topic_dict: The topics associated with the packages
+        in the input_stack+recommendation.
+        :return missing_packages: The list of packages unknown to the HPF model.
+        """
+        input_id_set = set()
+        missing_packages = []
+        for package_name in input_stack:
+            package_id = self.package_id_dict.get(package_name)
+            if package_id is None:
+                missing_packages.append(package_name)
+        else:
+            input_id_set.add(package_id)
+        # TODO: Check for known-unknown ratio before recommending
+        companion_recommendation, package_topic_dict = self.folding_in(
+            input_id_set)
+        return companion_recommendation, package_topic_dict, missing_packages
+
+    def match_manifest(self, input_id_set):
+        """Find a manifest list that matches user's input package list and return its index.
+
+        :param input_id_set: A set containing package ids of user's input package list.
+        :return manifest_id: The index of the matched manifest.
+        """
+        for manifest_id, dependency_set in self.manifest_id_dict.items():
+            if dependency_set == input_id_set:
+                break
+        else:
+            manifest_id = -1
+        return manifest_id
+
+    def folding_in(self, input_id_set):
+        """Folding in logic for prediction.
+
+        :param  input_id_set: A set containing package ids of user's input package list.
+        :return: Filter companion recommendations and their topics.
+        """
+        manifest_id = int(self.match_manifest(input_id_set))
+        if manifest_id == -1:
+            theta = []
+            with tf.Session() as sess:
+                epsilon = Gamma(self.a_c, self.a_c / self.b_c).eval()
+                theta_func = Gamma(self.a, epsilon).eval()
+                theta.append(epsilon * theta_func)
+                theta = np.array(theta * self.K)
+                result = Poisson(np.dot(theta, np.transpose(self.beta))).eval()
+            result = self.normalize_result(result)
+        else:
+            result = self.rating_matrix[manifest_id]
+        return self.filter_recommendation(result)
+
+    def normalize_result(self, result):
+        """Normalise the probability score of the resulting recommendation.
+
+        :param result: The Unnormalised recommendation result array.
+        :return result: The normalised recommendation result array.
+        """
+        maxn = result.max()
+        min_max = maxn - result.min()
+        for i in range(self.packages):
+            result[i] = (maxn - result[i]) / min_max
+        return result
+
+    def filter_recommendation(self, result):
```

Comment:
I would prefer if the sorting was taken care of using the numpy methods itself, since those are much faster. Take a look at `np.argsort` and `np.sort`, rather than pure Python sorting on lists.